app/models/user_annotation.rb

```diff
@@ -0,0 +1,291 @@
+        logger.info('created y in else')
+        #Create Cell Array
+        UserDataArray.create(name: 'text', array_type: 'cells', values: cell_name_arrays[i], cluster_name: cluster.name, array_index: i+1, user_id: self.user_id, cluster_group_id: self.cluster_group_id, study_id: self.study_id, user_annotation_id: id)
+        logger.info('created cells in else')
+      end
+    end
+
+    #If undefined happened then the annotation these data arrays belong to needs to have undefined in its values field
+    if undefined_happened
+      new_values = self.values
+      #Make sure not to duplicate undefined, but add it to the field values
+      if !new_values.include? 'Undefined'
+        new_values << 'Undefined'
+        self.update(values: values)
+      end
+    end
+
+  end
+
+  #If you created a selection at a lower level, and need to view it at higher levels, you must extrapolate that selection to those levels
+  #Extrapolating labels cells that were labeled at the lower level the same, and cells that weren't present in the lower level as 'Undefined'
+  def extrapolate(data, thresholds, max_length, cluster, annotation)
+    #full_thresh is the array of thresholds that data arrays have to be created at
+    #when threshold is nil, a data array at full data is created
+    #therefore, full thresh has nil by default to always create data arrays at no subsampling level
+    full_thresh = [nil]
+
+    #add thresholds to be created at, as long as the threshold isn't bigger than the size of full data, because that would be impossible
+    thresholds.each do |th|
+       if max_length > th
+         full_thresh << th
+       end
+    end
+
+    #create a data array at each threshold
+    full_thresh.each do |threshold|
+      create_array(cluster, threshold, annotation, data)
+    end
+
+  end
+
+  #If you created a selection at a higher level, and need to view it at lower levels, you must subsample that selection to those levels
+  #Subsampling labels cells that were labeled at the higher level the same in the lower level
+  def subsample(data, thresholds, cluster, annotation, max_length)
+    #create data arrays at this threshold as long as the threshold isn't bigger than the size of full data, because that would be impossible
+    thresholds.each do |threshold|
+      if threshold < max_length
+        create_array(cluster, threshold, annotation, data)
+      end
+    end
+  end
+
+  #figure out what level this annotation was created at
+  #if you update an annotation's 'Undefined' level, than it will be registered as created at full data
+  def subsampled_at
+    #get all the annotation arrays
+    annotations = user_data_arrays.find_by(array_type: 'annotations').to_a
+
+    #get the labels of this annotation
+    vals = values
```

Comment:
This association is ambiguous - is 'values' the attribute 'values' for the instance of UserAnnotation?  If so, this needs to be written as self.values.